<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
</head>

<body>

  <script>
    // Promise 是异步编程的一种解决方案，比传统的解决方案——回调函数和事件——更合理和更强大。它由社区最早提出和实现，ES6 将其写进了语言标准，统一了用法，
    // 原生提供了Promise对象。

    // 所谓Promise，简单说就是一个容器，里面保存着某个未来才会结束的事件（通常是一个异步操作）的结果。从语法上说，Promise 是一个对象，从它可以获取异步操作的消息。
    // Promise 提供统一的 API，各种异步操作都可以用同样的方法进行处理。

    // promise有以下特点:
    // 1.对象状态不受外界影响,Promise代表一个异步操作,有三种状态:pending(进行中),fulfilled(已成功),rejected(以失败),
    // 只有异步操作的结果才能决定当前是哪一种状态,其他任何操作都无法改变.

    // 2.一旦状态改变，就不会在变，Promise的状态改变只有两种可能：pending->fulfilled和pending->rejected。只要这两种状态发生了，状态就凝固了不会再变；
    // 如果状态发生改变,可以对Promise对象添加回调函数可以立即得到这个结果,这与事件(Event)不同,事件是错过了再去监听是得不到结果的

    // Promise也有一些缺点。首先，无法取消Promise，一旦新建它就会立即执行，无法中途取消。其次，如果不设置回调函数，Promise内部抛出的错误，
    // 不会反应到外部。第三，当处于pending状态时，无法得知目前进展到哪一个阶段（刚刚开始还是即将完成）

    // 如果某些事件不断地反复发生，一般来说，使用 Stream 模式是比部署Promise更好的选择。

    const promise = new Promise(function (resolve, rejecte) {
      if (true) {
        resolve(value)
      } else {
        reject(value)
      }
    })
    // Promise构造函数接受一个函数作为参数，该函数的两个参数分别是resolve和reject。它们是两个函数，由 JavaScript 引擎提供，不用自己部署。
    // resolve函数的作用是，将Promise对象的状态从“未完成”变为“成功”（即从 pending 变为 resolved），在异步操作成功时调用，并将异步操作的结果，作为参数传递出去；reject函数的作用是，
    // 将Promise对象的状态从“未完成”变为“失败”（即从 pending 变为 rejected），在异步操作失败时调用，并将异步操作报出的错误，作为参数传递出去。



    // Promise实例生成以后，可以用then方法分别指定resolved状态和rejected状态的回调函数。
    promise.then(function (value) {
      // success
    }, function (error) {
      // failure
    })
    // then方法可以接受两个回调函数作为参数。第一个回调函数是Promise对象的状态变为resolved时调用，
    // 第二个回调函数是Promise对象的状态变为rejected时调用。这两个函数都是可选的，不一定要提供。它们都接受Promise对象传出的值作为参数





    // promise例子
    function timeout (ms) {
      return new Promise((resolve, reject) => [
        setTimeout(resolve, ms, 'done')
      ])
    }
    timeout(1000).then((value) => {
      console.log(value);
    })

    // 异步加载图片例子
    function loadImage (url) {
      return new Promise((resolve, reject) => {

        const image = new Image()

        image.onload = function () {
          resolve(image)
        }
        image.onerror = function () {
          reject(new Error('COULD NOT LOAD IMAGE AT ' + URL))
        }
        image.src = url
      })
    }
    try {
      loadImage('https://www.google.com/imgres?imgurl=https%3A%2F%2Fblog.photoshelter.com%2Fwp-content%2Fuploads%2F2018%2F09%2Fgoogle-images-crop-1.png&tbnid=sa_k_yK2Dmp4jM&vet=12ahUKEwj-xP6DnJCCAxVgyMkDHZVBCQQQMygBegQIARB3..i&imgrefurl=https%3A%2F%2Fblog.photoshelter.com%2F2013%2F04%2Ffind-your-images-online-using-reverse-image-search-on-google%2F&docid=_5DPFNf3QmekFM&w=787&h=442&q=Image%20search&ved=2ahUKEwj-xP6DnJCCAxVgyMkDHZVBCQQQMygBegQIARB3')

    } catch (error) {
      console.log(error);
    }


    // 使用promise实现ajax
    const getJSON = function (url) {
      const promise = new Promise(function (resolve, reject) {
        const handler = function () {
          if (this.readyState !== 4) {
            return;
          }
          if (this.status === 200) {
            resolve(this.respose)
          } else {
            reject(new Error(this.statusText))
          }
        };
        const client = new XMLHttpRequest();
        client.open('GET', url)
        client.onreadystatechange = handler;
        client.responseType = "json";
        client.setRequestHeader("Accept", "application/json")
        client.send()
      });
      return promise;
    };
    try {
      getJSON("https://www.baidu.com").then(function (json) {
        console.log('contents:' + json);
      }, function (error) {
        console.error('出错了!');
      })
    } catch (error) {
      console.log(error);
    }


    const p1 = new Promise(function (resolve, reject) {
      setTimeout(() => reject(new Error('FAIL'), 3000))
    })
    const p2 = new Promise(function (resolve, reject) {
      setTimeout(() => resolve(p1), 3000)
    })
    p2.then(result => console.log(result)).catch(error => console.log(error))
    // p1是一个 Promise，3 秒之后变为rejected。p2的状态在 1 秒之后改变，resolve方法返回的是p1。由于p2返回的是另一个 Promise，
    // 导致p2自己的状态无效了，由p1的状态决定p2的状态。所以，后面的then语句都变成针对后者（p1）。又过了 2 秒，p1变为rejected，导致触发catch方法指定的回调函数。

    // 调用resolve或reject并不会终结 Promise 的参数函数的执行。
    new Promise(function (resolve, reject) {
      resolve(1)
      console.log(2);
    }).then(result => console.log(result))
    //2
    //1

    // 一般来说，调用resolve或reject以后，Promise 的使命就完成了，后继操作应该放到then方法里面，而不应该直接写在resolve或reject的后面。
    // 所以，最好在它们前面加上return语句，这样就不会有意外。
    // new Promise((resolve, reject) => {
    //   return resolve(1);
    //   // 后面的语句不会执行
    //   console.log(2);
    // })

    // Promise.prototype.catch()方法是.then(null, rejection)或.then(undefined, rejection)的别名，用于指定发生错误时的回调函数。
    getJSON('./POST.json').then((val) => console.log('fulfilled success')).catch((err) => console.log(err))

    // 一般来说，不要在then()方法里面定义 Reject 状态的回调函数（即then的第二个参数），总是使用catch方法。
    // bad
    promise.then((val) => console.log('success!'), (err) => console.log(err))
    // good 推荐用法
    promise.then((val) => console.log(val)).catch((err) => console.log(err))
    // 跟传统的try/catch代码块不同的是，如果没有使用catch()方法指定错误处理的回调函数，Promise 对象抛出的错误不会传递到外层代码，即不会有任何反应。
    // 一般总是建议，Promise 对象后面要跟catch()方法，这样可以处理 Promise 内部发生的错误。catch()方法返回的还是一个 Promise 对象，因此后面还可以接着调用then()方法。
    // catch后面可以再加catch捕获错误
    Promise.resolve().catch(function (error) { console.log('ohno error', error) }).then(function () {
      console.log('carry on');
    })


    // finally()方法用于指定不管 Promise 对象最后状态如何，都会执行的操作。该方法是 ES2018 引入标准的
    // 服务器请求例子
    // finally方法的回调函数不接受任何参数，这意味着没有办法知道，前面的 Promise 状态到底是fulfilled还是rejected。
    // 这表明，finally方法里面的操作，应该是与状态无关的，不依赖于 Promise 的执行结果。

    // const server = new Promise((resolve, reject) => resolve(1))
    // server.listen(port)
    //   .then(function () {

    //   }).finally(server.stop);

    // finally本质上是then方法的特例。

    // promise
    //   .finally(() => {
    //     // 语句
    //   });

    // // 等同于
    // promise
    //   .then(
    //     result => {
    //       // 语句
    //       return result;
    //     },
    //     error => {
    //       // 语句
    //       throw error;
    //     }
    //   );




    // Promise.all()方法用于将多个 Promise 实例，包装成一个新的 Promise 实例。

    // 生成一个Promise对象的数组
    const promises = [2, 3, 5, 7, 11, 13].map(function (id) {
      return getJSON('/post/' + id + ".json")
    })

    // 必须promise数组中所有都要fullfiled,promise.all才能fullfiled,只有一个rejected,就会catch
    Promise.all(promises).then(function (posts) {
      console.log('all success', posts);
    }).catch(function (reason) {
      console.log(reson);
    })

    const p5 = new Promise((resolve, reject) => {
      resolve('hello')
    }).then(result => result)
    const p6 = new Promise((resolve, reject) => {
      resolve('hello')
    }).then(result => result)
    // .catch(error => error)//如果这边catch了就会返回一个resolve对象,promise.all那边就抓不到错误
    const pArr = Promise.all([p5, p6]).then(result => console.log(result)).catch(e => console.log(e))






    // Promise.race()
    // Promise.race()方法同样是将多个 Promise 实例，包装成一个新的 Promise 实例。
    const pRaceArr = Promise.race([p5, p6]).then((resolve) => console.log(resolve)).catch(error => console.log(error))
    //只要p5和p6之间有一个实例先改变状态,p的状态也就跟着改变
    // 实例
    // 如果 5 秒之内fetch方法无法返回结果，变量p的状态就会变为rejected，从而触发catch方法指定的回调函数。
    const p = new Promise.race([
      fetch('https://www.baidu.com'),
      new Promise((resolve, reject) => setTimeout(reject(new Error('request timeout')), 5000))
    ]).then(result => console.log(result)).catch(err => console.log(err))




    // ES2020 引入了Promise.allSettled()方法，用来确定一组异步操作是否都结束了（不管成功或失败）。所以，它的名字叫做”Settled“，包含了”fulfilled“和”rejected“两种情况。
    // Promise.allSettled()方法接受一个数组作为参数，数组的每个成员都是一个 Promise 对象，并返回一个新的 Promise 对象。只有等到参数数组的所有 Promise 对象都发生状态变更
    // （不管是fulfilled还是rejected），返回的 Promise 对象才会发生状态变更。
    const promises1 = [
      fetch('https://www.baidu.com'),
      fetch('https://www.google.com'),
      fetch('https://www.linkin.com'),
    ]
    // await promises1.allSettled(promises)//等待所有异步请求执行完毕再进行下一步
    // 它的回调函数会接收到一个数组作为参数，该数组的每个成员对应前面数组的每个 Promise 对象。

    const resolve2 = Promise.resolve(666)
    const reject2 = Promise.reject(-1)

    const allSettledArr = Promise.allSettled([resolve2, reject2])
    console.log(allSettledArr);
    // [
//    { status: 'fulfilled', value: 42 },
//    { status: 'rejected', reason: -1 }
// ]
  </script>
</body>

</html>